src/legacy/bitstream.rs

```diff
@@ -0,0 +1,151 @@
+    bitpos_end: usize, /* Position of past-the-end bit. */
+}
+
+/// Initialize an input stream to present the n bytes from src as an LSB-first
+/// bitstream.
+impl<'a> BitStream<'a> {
+    pub fn new(src: &'a [u8], n: usize) -> Self {
+        Self {
+            src,
+            bitpos: 0,
+            bitpos_end: n * 8,
+        }
+    }
+
+    /// Get the next bits from the input stream. The number of bits returned is
+    /// between ISTREAM_MIN_BITS and 64, depending on the position in the stream, or
+    /// fewer if the end of stream is reached. The upper bits are zero-padded.
+    pub fn bits(&mut self) -> u64 {
+        let next = self.bitpos / 8;
+        debug_assert!(next < self.src.len(), "Cannot read past end of stream.");
+
+        let bits = if next + 8 <= self.src.len() {
+            // Common case: read 8 bytes in one go.
+            u64::from_le_bytes(self.src[next..next + 8].try_into().unwrap())
+        } else {
+            // Read the available bytes and zero-pad.
+            let mut bits = 0;
+            for i in 0..self.src.len() - next {
+                bits |= (self.src[next + i] as u64) << (i as u32 * 8);
+            }
+            bits
+        };
+
+        return bits >> (self.bitpos % 8);
+    }
+
+    /// Advance n bits in the bitstream if possible. Returns false if that many bits
+    /// are not available in the stream.
+    pub fn advance(&mut self, n: u8) -> std::io::Result<()> {
+        debug_assert!(self.bitpos <= self.bitpos_end);
+
+        if self.bitpos_end - self.bitpos < n as usize {
+            return Err(io::Error::new(
+                io::ErrorKind::UnexpectedEof,
+                "End of stream",
+            ));
+        }
+
+        self.bitpos += n as usize;
+        Ok(())
+    }
+
+    /// Align the input stream to the next 8-bit boundary and return a pointer to
+    /// that byte, which may be the past-the-end-of-stream byte.
+    pub fn _byte_align(&mut self) -> usize {
+        debug_assert!(self.bitpos <= self.bitpos_end, "Not past end of stream.");
+        self.bitpos = 8 * (self.bitpos / 8);
+        debug_assert!(self.bitpos <= self.bitpos_end, "Not past end of stream.");
+        return self.bitpos / 8;
+    }
+
+    pub fn bytes_read(&self) -> usize {
+        (self.bitpos + 7) / 8
+    }
+    
+    pub fn read_next_bits(&mut self, code_size: u8) -> std::io::Result<u64> {
+        let b = self.bits();
+        self.advance(code_size)?;
+        Ok(lsb(b, code_size) )
+    }
+}
+
+pub const ISTREAM_MIN_BITS: usize = 64 - 7;
+
+#[cfg(test)]
+mod tests {
+    use crate::legacy::bitstream::{lsb, reverse_lsb};
+
+    #[test]
+    fn test_reverse16() {
+        assert_eq!(reverse_lsb(0x0000, 1), 0x0);
+        assert_eq!(reverse_lsb(0xffff, 1), 0x1);
+        assert_eq!(reverse_lsb(0x0000, 16), 0x0);
+        assert_eq!(reverse_lsb(0xffff, 16), 0xffff);
+        // 0001 0010 0011 0100 -> 0010 1100 0100 1000
+        assert_eq!(reverse_lsb(0x1234, 16), 0x2c48);
```

Comment:
Use binary literals on this line and line 108.

Reply:
switched to bitstream-io.